contributors/design-proposals/lifecycle-hooks.md

```diff
@@ -0,0 +1,152 @@
+Lifecycle hooks should allow users to better customize updates (rolling/recreate) without having to write their own controllers. 
+
+## Lifecycle Hooks
+Lifecycle hook is a Job that will get triggered by the update reaching certain progress point (like 50%). 
+The progress point is determined by the ratio of new.availableReplicas to the declared replicas or by explicitly stating the number of new.availableReplicas. 
+The API is open for other types of triggering conditions in the future. 
+
+### Previous Implementations
+Lifecycle hooks are already implemented in OpenShift but we want to enhance them and implement them natively in Kubernetes.
+ - [OpenShift proposal](https://github.com/openshift/origin/blob/master/docs/proposals/post-deployment-hooks.md)
+ - [OpenShift docs](https://docs.openshift.org/latest/dev_guide/deployments/deployment_strategies.html#lifecycle-hooks)
+
+### Previous Kubernetes Proposals
+ - https://github.com/kubernetes/kubernetes/pull/33545
+
+### Use Cases
+The most common use case for lifecycle hooks is to have pre, mid and/or post hook. 
+It is mostly used to run some kind of acceptance check in the middle and/or at the end to fully verify the update is working as expected and rollback if it isn't. 
+The acceptance check may be time consuming and more thorough than what readiness and liveness probes are intended for. 
+You can also notify external services from them, migrate database in the middle of an update, send messages to IRC channel or do anything else.
+
+[A short demo](https://youtu.be/GVNTm_K43iI) simulating lifecycle hooks using auto-pausing that has been presented at SIG-Apps meeting on August 21, 2017.
+
+
+## API Objects
+### New Objects
+```go
+type HookTriggerType string
+
+const (
+	ProgressAvailableHookTriggerType HookTriggerType = "ProgressAvailable"
+)
+
+type ProgressAvailableParams struct {
+    // ProgressPoint specifies the point during update when the hook should be triggered.
+    // Accepts both the number of new pods available or a percentage value representing the ratio
+    // of new.availableReplicas to the declared replicas. In case of getting a percentage
+    // it will try to reach the exact or the closest possible point right after it,
+    // trigger the job, wait for it to complete and then continue. 
+    // If such situation shall occur that two different ProgressPoints should be reached at
+    // the same time, all the hooks for an earlier ProgressPoint will be ran (and finished)
+    // before any later one.
+    ProgressPoint intstr.IntOrString
+}
+
+type HookTrigger struct {
+    // Type of hook trigger. Can be ProgressAvailable. Default is ProgressAvailable.
+    Type HookTriggerType
+    
+    // ProgressAvailable config params. Present only if type = "ProgressAvailable".
+    ProgressAvailableParams *ProgressAvailableParams
```

Comment:
@apelisse Do we have conventions for unions yet?

Reply:
Nothing formal yet, but new APIs that have "oneof" structs should try to:
- Have a struct that only contains:
  - a discriminator as a string,
  - optional (with `// +optional`), omitempty, pointer fields,
- That structure must not be inlined,
- Values for the discriminator must match corresponding field names (modulo capitalized first letter).